curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -960,62 +1037,75 @@ void ChunkCacheManager::ReleaseCache(S3ClientAdaptorImpl* s3ClientAdaptor) {
                       << ",data chunkpos:" << iter->second->GetChunkPos();
-            if (dataWCacheMap_.empty()) {
-                WriteLockGuard writeLockGuard(rwLockWrite_);
-                dataWCacheMap_.swap(tmp);
-            } else {
-                // todo
-                s3ClientAdaptor_->GetFsCacheManager()->DataCacheNumFetchSub(
-                    size);
-            }
             return ret;
         }
         if (ret == CURVEFS_ERROR::OK) {
             AddReadDataCache(iter->second);
+            if (!iter->second->IsDirty()) {
+                WriteLockGuard writeLockGuard(rwLockWrite_);
+                if (dataWCacheMap_.count(iter->first)) {
+                    dataWCacheMap_.erase(iter->first);
+                    s3ClientAdaptor_->GetFsCacheManager()->DataCacheNumFetchSub(
+                        1);
+                    LOG(INFO) << "chunk flush DataCacheByteDec1 len:"
+                              << iter->second->GetLen() << "force:" << force;
+                    s3ClientAdaptor_->GetFsCacheManager()->DataCacheByteDec(
+                        iter->second->GetLen());
+                    if (!s3ClientAdaptor_->GetFsCacheManager()
+                             ->WriteCacheIsFull()) {
+                        LOG(INFO) << "write cache is not full, signal wait.";
+                        s3ClientAdaptor_->GetFsCacheManager()->FlushSignal();
+                    }
+                }
+            } else {
+                LOG(INFO) << "data cache is dirty.";
+            }
         }
-        s3ClientAdaptor_->GetFsCacheManager()->DataCacheNumFetchSub(1);
     }
 
     return CURVEFS_ERROR::OK;
 }
 
 void DataCache::Write(uint64_t chunkPos, uint64_t len, const char* data,
                       const std::vector<DataCachePtr>& mergeDataCacheVer) {
-    uint64_t total_size = 0;
-
+    uint64_t totalSize = 0;
+    uint64_t addByte = 0;
     LOG(INFO) << "DataCache Write() chunkPos:" << chunkPos << ",len:" << len
               << ",dataCache's chunkPos:" << chunkPos_
               << ",dataCache's len:" << len_;
     curve::common::LockGuard lg(mtx_);
+    dirty_.exchange(true, std::memory_order_acq_rel);
     if (chunkPos <= chunkPos_) {
         /*
             ------       DataCache
          -------         WriteData
         */
         if (chunkPos + len <= chunkPos_ + len_) {
-            total_size = chunkPos_ + len_ - chunkPos;
-            char* newDatabuf = new char[total_size];
+            totalSize = chunkPos_ + len_ - chunkPos;
+            addByte = totalSize - len_;
+            s3ClientAdaptor_->GetFsCacheManager()->DataCacheByteInc(addByte);
+            char* newDatabuf = new char[totalSize];
             memcpy(newDatabuf, data, len);
-            memcpy(newDatabuf + len, data_, total_size - len);
-            char* tmp = data_;
-            data_ = newDatabuf;
-            delete tmp;
-            len_ = total_size;
+            memcpy(newDatabuf + len, data_, totalSize - len);
```

Comment:
possible optimization for alleviating memory copy is using a policy like string_view. but currently, this will introduce more modifications.